---
title: "Analysis"
author: "rbabaei"
date: "6/25/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE, cache = TRUE)
```

```{r packages, include=FALSE}
required_packages = c("curatedTCGAData", "survival", "survminer", "GenomicRanges", "foreach", "doParallel",
                      "pheatmap", "ggplot2", "tidyverse", "tictoc", "missForest", "NMF", "enrichR")
suppressPackageStartupMessages(lapply(required_packages, require, character.only =T))
```

## Synopsis

The multi-omics data from breast invasive carcinoma project of The Cancer Genome Atlas (TCGA) is downloaded and analyzed. A subset of the data is chosen according to the clinical information of the patients. A clear workflow has been used to be reproducible. The analysis has been concluded by biological interpretation, which is relevant to the analyzed subset. The original data are accessible at: https://portal.gdc.cancer.gov/projects/TCGA-BRCA  


## Data download and cleaning

We'll use the curatedTCGA package to download the omics data from TCGA data portal. We then will have a closer look at data structure, integration and some exploratory analysis.

```{r data}
curatedTCGAData(diseaseCode = "BRCA")
BRCAmae <- curatedTCGAData("BRCA", c("GISTIC_ThresholdedByGene","Mutation", "RNASeq2GeneNorm", "RPPAArray"),
                           dry.run = FALSE)
```

From the listed available data at TCGA, we've selected the copy number variation, gene expression, mutation, and the reverse phase protein array. 

```{r BRCA}
dim(colData(BRCAmae))
table(BRCAmae$histological_type)
```

As we see through meta data provided by MultiAssayExpreiment file, there are 1098 patients studied in this project, and their clinical data are categorized in 2684 variables. If look at histological types of the diagnosed patients, we see that the infiltrating ductal and lobular carcinomas are the most prevalent types. We will later select only these two types  for further analysis. But first, let's look at the data as a whole.


```{r upset, fig.align="center"}
upsetSamples(BRCAmae)
```

780 samples are including all assays, and only three are missing the gene expression assay.
With the help of kaplan-meier plot, we can also explore the survival probablity of the patients by the time, based on their nodal stage. The number at risk at each time span is summarized in the table below.


```{r kaplan-meier, fig.align="center"}
BRCAmaesurv <- BRCAmae[, complete.cases(BRCAmae$days_to_death, BRCAmae$vital_status), ]


fit <- survfit(Surv(days_to_death, vital_status)~Node, data = colData(BRCAmaesurv))
ggsurvplot(fit, data = colData(BRCAmaesurv), risk.table = T)
```

In this step, we will clean, reshape, and normalize the data if needed for further analysis.

```{r dataPreparation}
## col_data for graph annotation
dfcolData <- data.frame(matrix(nrow = nrow(colData(BRCAmae))))
for(i in 1:ncol(colData(BRCAmae))){dfcolData <- cbind(dfcolData, BRCAmae@colData@listData[i])}
dfcolData <- dfcolData[,-1]
rownames(dfcolData) <- BRCAmae@colData@rownames
colnames(dfcolData) <- gsub("\\-|\\_", ".", colnames(dfcolData))
rownames(dfcolData) <- gsub("\\-|\\_", ".", rownames(dfcolData))

toBeRemoved <- which(data.frame(map(dfcolData, ~sum(is.na(.))))> nrow(dfcolData)*0.3)# just to reduce the dimension, not for imputing
dfcolData <- dfcolData[,-(toBeRemoved)]

# select patients with only ductal or lobular carcinoma
 dfcolData <- dfcolData %>% filter(histological.type == "infiltrating ductal carcinoma" | histological.type == "infiltrating lobular carcinoma")


######
listAssays <- assays(BRCAmae)

#######
## RNA-Seq
RNASeq <- listAssays[[3]]
colnames(RNASeq) <- gsub("(\\-[[:alnum:]]+\\-[[:alnum:]]+\\-[[:alnum:]]+\\-[0-9]+)$", "", colnames(RNASeq))
# log transformation
RNASeq <- log2(RNASeq +1)

dim(RNASeq)
length(unique(colnames(RNASeq)))# there are duplicated variables

RNASeq <- RNASeq[,unique(colnames(RNASeq))] # remove duplicated variables
colnames(RNASeq) <- gsub("\\-|\\_", ".", colnames(RNASeq))

# select patients with only ductal or lobular carcinoma
 RNASeq <- RNASeq[,intersect(colnames(RNASeq), rownames(dfcolData))]
# dim(RNASeq)

RNASeq <- t(RNASeq)


## remove low variable genes
SDs <- apply(RNASeq,2,sd)
orderSD <- order(SDs, decreasing = T)[1:2000]
RNASeq <- RNASeq[,orderSD]
dim(RNASeq)


##copy number
copyNumber <- listAssays[[1]]
colnames(copyNumber) <- gsub("(\\-[[:alnum:]]+\\-[[:alnum:]]+\\-[[:alnum:]]+\\-[0-9]+)$", "", colnames(copyNumber))
colnames(copyNumber) <- gsub("\\-|\\_", ".", colnames(copyNumber))

# select patients with only ductal or lobular carcinoma
 copyNumber <- copyNumber[,intersect(colnames(copyNumber), rownames(dfcolData))]
 dim(copyNumber)
length(unique(colnames(copyNumber))) ## there are unique variables
sum(is.na(copyNumber)) # there is no missing value
copyNumber <- t(copyNumber)
copyNumber <- copyNumber[,intersect(colnames(copyNumber), colnames(RNASeq))] # relative to RNASeq
dim(copyNumber)

## mutation
Mutation <- listAssays[[2]]
colnames(Mutation) <- gsub("(\\-[[:alnum:]]+\\-[[:alnum:]]+\\-[[:alnum:]]+\\-[0-9]+)$", "", colnames(Mutation))

# reshape the sparse matrix and set binary values for mutations
Mutationdf <- as.data.frame(Mutation)
tmp <- data.frame(GeneSymbol = rep(NA, nrow(Mutationdf)))
for(i in 1:ncol(Mutationdf)){
  tmp$GeneSymbol <- coalesce(Mutationdf[,i], tmp$GeneSymbol)
  Mutationdf$GeneSymbol <- tmp$GeneSymbol
}

sum(is.na(Mutationdf$GeneSymbol))# no missing value

tmp <- data.frame(GeneSymbol = unique(Mutationdf$GeneSymbol))# select unique gene symbols
sum(is.na(tmp$GeneSymbol))# no NA


cl <- makeCluster(3)
registerDoParallel(cl)

tmp_data <- foreach(i = c(1:993), j = c(1:993))%dopar%{
  tmp[,paste0("V", i)] <- ifelse(tmp$GeneSymbol%in%Mutationdf[,j], 1, 0)
}
stopCluster(cl)

tmp_df <- data.frame(matrix(unlist(tmp_data), ncol = length(tmp_data)))# convert to data frame
colnames(tmp_df) <- colnames(Mutationdf[,1:993])
rownames(tmp_df) <- tmp$GeneSymbol
Mutationdf <- tmp_df
colnames(Mutationdf) <- gsub("\\-|\\_", ".", colnames(Mutationdf))

# select patients with only ductal or lobular carcinoma
 Mutationdf <- Mutationdf[,intersect(colnames(Mutationdf), rownames(dfcolData))]
dim(Mutationdf)

length(unique(colnames(Mutationdf)))# unique variables
rm(tmp, tmp_data, tmp_df)
Mutation <- t(as.matrix(Mutationdf))

# filter out genes with too few mutations (less than 5%)
mut.rate <- apply(Mutation, 2, mean)
Mutation <- Mutation[,which(mut.rate>0.02)]
dim(Mutation)


## PP
RPPA <- listAssays[[4]]
dfRPPA <- data.frame(matrix(nrow = 226))
for(i in 1:937){dfRPPA <- cbind(dfRPPA, RPPA@listData[i])}
dfRPPA <- dfRPPA[,-1]
rownames(dfRPPA) <- RPPA@rownames
colnames(dfRPPA) <- gsub("(\\-[[:alnum:]]+\\-[[:alnum:]]+\\-[[:alnum:]]+\\-[0-9]+)$", "", colnames(dfRPPA))
rownames(dfRPPA) <- gsub("\\-|\\_", ".", rownames(dfRPPA))

length(unique(colnames(dfRPPA)))# there are duplicated variables
dfRPPA <- dfRPPA[,unique(colnames(dfRPPA))]
colnames(dfRPPA) <- gsub("\\-|\\_", ".", colnames(dfRPPA))

# select patients with only ductal or lobular carcinoma
 dfRPPA <- dfRPPA[,intersect(colnames(dfRPPA), rownames(dfcolData))]
mRPPA <- t(as.matrix(dfRPPA))
dim(mRPPA)
sum(is.na(mRPPA))# there are missing values
toBeRemoved <- which(data.frame(map(as.data.frame(mRPPA), ~sum(is.na(.))))> nrow(mRPPA)/2) # three variables with more than 50% missing values, to be removed

mRPPA <- mRPPA[,-c(toBeRemoved)]
# impute the rest missing values
set.seed(4578)
registerDoParallel(cores = 3)
impute_forest <- missForest(data.frame(mRPPA), maxiter = 10, ntree = 100, parallelize = "forests")
stopImplicitCluster()

impute_forest$OOBerror # 90% accuracy
mRPPA <- as.matrix(impute_forest$ximp)
sum(is.na(mRPPA))# no missing value



rm(BRCAmae, cl, dfRPPA, listAssays, Mutationdf, RPPA, mut.rate, SDs)

#### list of omics with same patients
dataList <- list(CN =copyNumber,MU= Mutation,RS= RNASeq,PP= mRPPA)
intersectID <- Reduce(intersect, lapply(dataList, rownames))
dataList <- lapply(dataList, function(x) x[intersectID,]) # data is selected for 705 patients

## filter colData by shared rows
dfcolData <- dfcolData[intersect(rownames(dfcolData), rownames(dataList$CN)),]

```

We ened up with 705 patients. 
Most probably there could be a relation between the copy number and the level of gene expression. Let's find it out.
Below the top 10 genes with the highest correlation between copy number and gene expression are listed, and the correlation of the top gene is plotted.  

```{r corr, fig.align="center"}
tic("Correlation")
corr <- cor(dataList$CN, dataList$RS)
toc()

# top ten correlated genes
top10 <- order(diag(corr), decreasing = T)[1:10]
k <- arrayInd(top10, dim(corr))
topGenes <- rownames(corr)[k[,1]]
topGenes
# boxplot of the top gene
dfTop <- data.frame(RNASeq = dataList$RS[,topGenes[1]], CopyNumber = dataList$CN[,topGenes[1]])
colnames(dfTop) <- c("RNASeq", "CopyNumber")
boxplot(RNASeq~CopyNumber, data = dfTop, xlab = "GISTIC Relative Copy Number Call", ylab = "RNA-Seq log FC", main = topGenes[1])
```


## Joint Non-negative Matrix Factorization (NMF)

We are going to use the NMF as latent variable model for multi omics integration. In our data set, copy number and reverse phase protein array have negative values. To be abale to use the NMF algorithm, we should first convert these data into a non-negative form. To do that, we will split each feature of them into two features, one for the non-negative values and another one for the absolute value of the negative values. We will then use the long and wide formats of the NMF produced matrices to go deeper into the features and clinical characteristics. 

```{r nmf}
# Feature-normalize the data
CN.featnorm <- t(dataList$CN)/rowSums(t(abs(dataList$CN)))# take abs to avoid rowsums of zero
MU.featnorm <- t(dataList$MU)/rowSums(t(dataList$MU))
RS.featnorm <- t(dataList$RS)/rowSums(t(dataList$RS))
PP.featnorm <- t(dataList$PP)/rowSums(t(abs(dataList$PP)))# take abs to avoid rowsums of zero

# Normalize by each omics type's frobenius norm
CN.featnorm.frobnorm <- CN.featnorm/norm(as.matrix(CN.featnorm), type = "F")
MU.featnorm.frobnorm <- MU.featnorm/norm(as.matrix(MU.featnorm), type = "F")
RS.featnorm.frobnorm <- RS.featnorm/norm(as.matrix(RS.featnorm), type = "F")
PP.featnorm.frobnorm <- PP.featnorm/norm(as.matrix(PP.featnorm), type = "F")

# Split the features of the CN and PP matrix into two non-negative features each
split_neg_columns <- function(x) {
  new_cols <- list()
  for(i in seq_len(dim(x)[2])) {
    new_cols[[paste0(colnames(x)[i],'+')]] <- sapply(X = x[,i], 
                                                     function(x) max(0,x))
    new_cols[[paste0(colnames(x)[i],'-')]] <- sapply(X = -x[,i],
                                                     function(x) max(0,x))
  }
  new_cols
  return(do.call(cbind, new_cols))
}

CN.nonneg <- t(split_neg_columns(t(CN.featnorm.frobnorm)))
PP.nonneg <- t(split_neg_columns(t(PP.featnorm.frobnorm)))


r.nmf <- nmf(t(rbind(CN.nonneg, MU.featnorm.frobnorm, RS.featnorm.frobnorm, PP.nonneg)), 2, method = "Frobenius")
nmf.h <- basis(r.nmf)
nmf.w <- NMF::coef(r.nmf)
```

We will try to cluster the patients using two components created by NMF.

```{r nmfComponents, fig.align="center"}
nmf_df <- as.data.frame(nmf.h)
colnames(nmf_df) <- c("dim1", "dim2")
nmf_df$subtypes <- factor(dfcolData[rownames(nmf_df),]$histological.type)
nmf_df$ER <- factor(dfcolData[rownames(nmf_df),]$patient.breast.carcinoma.estrogen.receptor.status)

table(nmf_df$subtypes)

ggplot(nmf_df, aes(x = dim1, y = dim2, color = subtypes, shape = ER)) + geom_point() + ggtitle("2-component NMF for multi-omics integratoin")
```

We were able to separate the patients with infiltrating ductal carcinoma from the ones with infiltrating lobular carcinoma. Interestingly, most of the patients with lobular carcinoma are estrogen receptor pisitive, whereas lots of the patients with ductal carcinoma are estrogen receptor negative.

```{r pheatmap, fig.align="center"} 
anno_col <- data.frame(type = as.factor(dfcolData$histological.type))
rownames(anno_col) <- rownames(dfcolData)
pheatmap(t(nmf_df[,1:2]),annotation_col = anno_col, show_colnames=FALSE, main="Heatmap of 2-component NMF")
```

The figure above shows that most tumors are driven mainly by one of the factors. Based on this, we will assign each tumor a cluster label by its dominant factor. 

```{r oneHot, fig.align="center"}
nmf.cluster <- max.col(nmf.h)
names(nmf.cluster) <- rownames(nmf.h)
anno_nmf_cl <- data.frame(nmf.cluster = factor(nmf.cluster), subtypes = factor(dfcolData[rownames(nmf.h),]$histological.type))

pheatmap(t(nmf.h[order(nmf.cluster),]), cluster_cols = F, cluster_rows = F, annotation_col = anno_nmf_cl,
         show_colnames = F, main = "Joint NMF factors with clusters and subtypes")
```

Using one-hot clustering with Joint NMF, two clusters showed up in the data, which are fairly relevant to the histological subtypes of tumors.

## Biological Interpretation

By looking at the coefficients of the latent factors, we can find out which multi omics features are co-regulated.

```{r coef, fig.align="center"}
nmfw <- t(nmf.w)
data_anno <- data.frame(omics=c(rep('expression',dim(dataList$RS)[2]), rep('mut',dim(dataList$MU)[2]),
                                rep('cnv',dim(CN.nonneg)[1]), rep("RPP", dim(PP.nonneg)[1])))
rownames(data_anno) <- c(rownames(t(dataList$RS)), paste0("mut:", rownames(t(dataList$MU))),paste0("CN:", rownames(CN.nonneg)), rownames(PP.nonneg))
rownames(nmfw) <- rownames(data_anno)

pheatmap::pheatmap(nmfw,
                   cluster_cols = FALSE,
                   annotation_row = data_anno,
                   main="NMF coefficients",
                   clustering_distance_rows = "manhattan",
                   fontsize_row = 1)
```

```{r featureInspection}
topExp <- data.frame(nmfw[colnames(dataList$RS),])
colnames(topExp) <- c("F1","F2")
knitr::kable(topExp[order(topExp$F1, decreasing = T),][1:3,])
knitr::kable(topExp[order(topExp$F2, decreasing = T),][1:3,])



topCN <- data.frame(nmfw[paste0("CN:", rownames(CN.nonneg)),])
colnames(topCN) <- c("F1","F2")
knitr::kable(topCN[order(topCN$F1, decreasing = T),][1:3,])
knitr::kable(topCN[order(topCN$F2, decreasing = T),][1:3,])



topMU <- data.frame(nmfw[paste0("mut:", colnames(dataList$MU)),])
colnames(topMU) <- c("F1","F2")
knitr::kable(topMU[order(topMU$F1, decreasing = T),][1:3,])
knitr::kable(topMU[order(topMU$F2, decreasing = T),][1:3,])


topPP <- data.frame(nmfw[rownames(PP.nonneg),])
colnames(topPP) <- c("F1","F2")
knitr::kable(topPP[order(topPP$F1, decreasing = T),][1:3,])
knitr::kable(topPP[order(topPP$F2, decreasing = T),][1:3,])
```

By inspecting the features, we see that gene expression and mutation dominate only one factor, whereas copy number and phase protein dominate both factors.
As shown above, the first factor is associated with high expression of CACNA1H, CACNA2D1, and C21orf15 genes, loss of KRT1, RDH16, and NXPH4 copy numbers, frequent mutation of MGA, SYNE1, and LRBA, and higher level of Cyclin.E1, FOXM1, and ASNS proteins. Second factor is associated with the gain in copy number of KLK11, DPYSL, and ASCL2, and decrease in Caspase3, SCD, and MyosinIIa proteins.


## Enrichment Analysis

To show that the genes associated with a factor (here factor one) significantly overlap the genes in a biological process, we will apply the hypergeometric enrichment test (Fisher's one-sided exact test).

```{r enrichment, fig.align="center"}
# gene expression affect only the first factor

genes.factor <- rownames(topExp[order(topExp$F1, decreasing = T),][1:100,])# analysis for top 100 genes
go.factor <- enrichr(genes.factor, databases = c("GO_Biological_Process_2018"))$GO_Biological_Process_2018

ggplot(head(go.factor), aes(y=-log10(P.value), x = Term))+
  geom_bar(stat = "identity") + coord_flip() + ggtitle("Top GO-Terms\nassociated with\nNMF factor")

```

We confirmed that the expression of the genes associated with the first factor are significantly overlapping the biological pathways, such as acute inflamatory response, regulation of protein phosphorylation, and the regulation of chemotaxis.


## Session Info

```{r sessionInfo, echo=FALSE}
sessionInfo()
```






